src/models/BrowserStyleSheet.js

```diff
@@ -116,36 +105,285 @@ class BrowserTag implements Tag {
+      if (!this.ready) this.replaceElement()
+
+      if (
+        process.env.NODE_ENV !== 'production' &&
+        this.components[componentId]
+      ) {
+        throw new Error(`Trying to add Component '${componentId}' twice!`)
+      }
+
+      this.components[componentId] = {
+        componentIndex: this.componentSizes.length,
+        css: '',
+      }
+
+      this.componentSizes.push(0)
+      this.size += 1
+    }
+
+    inject(componentId: string, cssRules: Array<string>) {
+      if (!this.ready) this.replaceElement()
+
+      const comp = this.components[componentId]
+      if (process.env.NODE_ENV !== 'production' && !comp) {
+        throw new Error(
+          'Must add a new component before you can inject css into it'
+        )
+      }
+
+      const cssRulesSize = cssRules.length
+      const sheet = sheetForTag(this.el)
+      const { componentIndex } = comp
+
+      // Determine start index for injection
+      const insertIndex = sizeUpToComponentIndex(
+        this.componentSizes,
+        componentIndex
+      )
+
+      // Inject each rule shifting index forward for each one (in-order injection)
+      let injectedRules = 0
+      for (let i = 0; i < cssRulesSize; i += 1) {
+        const cssRule = cssRules[i]
+        if (safeInsertRule(sheet, cssRule, insertIndex + injectedRules)) {
+          comp.css += ` ${cssRule}`
+          injectedRules += 1
+        }
+      }
+
+      // Update number of rules for component
+      this.componentSizes[componentIndex] += injectedRules
+    }
+
+    toRawCSS() {
+      return '' // NOTE: Unsupported in production mode (SpeedyBrowserTag)
+    }
+
+    toHTML() {
+      return '' // NOTE: Unsupported in production mode (SpeedyBrowserTag)
+    }
+  }
+} else {
+  BrowserTag = class TextNodeBrowserTag extends BaseBrowserTag implements Tag {
+    isLocal: boolean
+    components: { [string]: Object }
+    size: number
+    el: HTMLStyleElement
+    ready: boolean
+
+    constructor(
+      el: HTMLStyleElement,
+      isLocal: boolean,
+      existingSource: string = ''
+    ) {
+      super()
 
-    // Build up our replacement style tag
-    const newEl = this.el.cloneNode()
-    newEl.appendChild(document.createTextNode('\n'))
+      const nonce = getNonce()
+      if (nonce !== null) {
+        el.setAttribute('nonce', nonce)
+      }
+
+      const extractedComps = extractCompsFromCSS(existingSource)
 
-    Object.keys(this.components).forEach(key => {
-      const comp = this.components[key]
+      this.el = el
+      this.isLocal = isLocal
+      this.ready = false
+      this.size = extractedComps.length
+      this.components = extractedComps.reduce((acc, obj) => {
+        acc[obj.componentId] = obj // eslint-disable-line no-param-reassign
+        return acc
+      }, {})
+    }
 
-      // eslint-disable-next-line no-param-reassign
-      comp.textNode = document.createTextNode(comp.cssFromDOM)
-      newEl.appendChild(comp.textNode)
-    })
+    getComponentIds() {
```

Comment:
does this need to be re-implemented since it's on `BaseBrowserTag`?

Reply:
Ah no, the rebase messed it up; good catch